### PR TITLE
Deploying laravel to kubernetes v2

### PR DESCRIPTION
**Pull request by keithmifsud**
@danielepolencic I have re-written the post. As previously discussed, I added the Ingress and removed the explanation of terms. I changed most of the content and taken into account all the comment from #32 .

Also, I got 100% score on Grammarly when ignoring the warnings on repetitive words which cannot be swapped. (Such as Seeds instead of Pods) 😄 

Please review. I think this is ready to be merged once you get the illustration or image. Please also let me know if there are any categories you'd like this filed under.

### CODE REVIEW

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+then you can deploy the container image:
+
+```bash
+kubectl run laravel-kubernetes-demo --image=yourname/laravel-kubernetes-demo 
+--port=8181 --image-pull-policy=IfNotPresent
+```
+The above command tells Kubectl to run our demo application from the Docker image while making port 8181 available for listening. The last parameter of the command simply asks Kubectl to not pull the image from a registry such as Docker Hub if it exists locally which in this case it does. Do note that you still need to be logged on to Docker's so that Kubectl can check if the image is up to date.
+
+You can check that a Pod is created for our application by running:
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: for your application?

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+or to acquire the dashboard's URL address:
+
+```bash
+minikube dashboard --url=true
+```
+
+__Exposing a service__
+
+So far you have created a deployment which is running the application's container. A pod running in the cluster has a dynamic IP. If you route the traffic directly to it using the IP, you may still need to update the routing table every time the pod is restarted. In fact, on every deployment or container restart, a new IP is assigned to the pod. To avoid managing IP addresses manually, you need to use a Service. The service acts as an IP and serves as a load balancer for a set of pods. So even if the IP address of a pod changes, the service is always pointing to it. And since the Service always has the same IP, you won't need to update anything manually.
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: The service acts as an IP and serves as a load balancer for a set of pods.

I think we could have:

The service acts as a load balancer for a set of pods.

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+or, launch the application directly in the browser:
+
+```bash
+minikube service laravel-kubernetes-demo
+```
+
+__Scaling__
+
+And that is it. You have successfully deployed the application in Kubernetes. It's exciting. But what's the point of doing all of this? Well, you only have one deployment with a single Pod running, provisioned to a Node with the exposed web service. Let's replicate this pod to two more pods acting as load balancers to the web traffic.
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: Let's replicate this pod to two more pods acting as load balancers to the web traffic.

Not sure what you meant with "acting as load balancers to the web traffic".

I think you could have: "Let's scale this deployment to two instances of the application"

Also note that later on you scale your replicas to 3. Not sure which one you wish to keep.

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+kubectl get deployment
+
+NAME                      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
+laravel-kubernetes-demo   3         3         3            3           59m
+```
+
+You can also see this in the Dashboard under Pods or in the Service detail screen.
+
+Now the application's deployment services are running on three 
+separate, replicated Pods.
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: I think this is an opportunity to explain why Kubernetes is cool. We could add:

Imagine your application becoming even more popular. Thousands of visitors are flocking to your website. In the past you may have been busy launching more scripts to create more instances of your application. In Kubernetes you can scale to ten instances in a snap:

```bash
kubectl scale --replicas=10 deployment/laravel-kubernetes-demo
deployment "laravel-kubernetes-demo" scaled
```

You can see how convenient it is to use Kubernetes to scale your website.

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+__Ingress__
+
+You've already achieved great things, you deployed the application and scaled the deployment. You have already seen the running application in the browser when pointed to the cluster's (Minikube) IP address and node's port number. Now, you will see how to access the application through an assigned URL as you would do when deploying to the cloud.
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: ^ this I like a lot. very good intro 👏

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+Typically a Kubernetes cluster is firewalled from the internet, and the same is true for the Minikube cluster. To use a URL, in Kubernetes, you need Ingress. An Ingress is a set of rules to allow inbound connections to reach a Kubernetes cluster. The Ingress is necessary because, in Kubernetes, resources such as Pods only have IP addresses only routable by and within the cluster. Meaning that they are not accessible or reachable to and from the world outside.
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: Not sure what you mean with this: "Typically a Kubernetes cluster is firewalled from the internet, and the same is true for the Minikube cluster."

But I like the following: "To use a URL, in Kubernetes, you need **an** Ingress."

Maybe we just rephrase or drop the first sentence?

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+  rules:
+  - host: laravel-kubernetes.demo
+  - http:
+      paths:
+      - path: /
+        backend:
+          serviceName: laravel-kubernetes-demo
+          servicePort: 8181
+```
+Among the basic content you would expect from a Kubernetes resource file, this file defines a set of rules for the Kubernetes firewall to follow for inbound traffic. The `laravel-kubernetes.demo` URL will point to the Pod where the application is running, as previously labelled `laravel-kubernetes-demo` on port 8181.
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: I don't think the ingress is a firewall. I think the right word should be reverse proxy. I'd rephrase as:

Among the basic content you would expect from a Kubernetes resource file, this file defines a set of rules to follow to route inbound traffic.

Also the following is not striclty true:

The `laravel-kubernetes.demo` URL will point to the Pod where the application is running, as previously labelled `laravel-kubernetes-demo` on port 8181.

You're not pointing to a Pod, but to the service. So it's Ingress -> Service -> Pod

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+
+The Ingress resource is useless without an Ingress controller so you will need to create a new controller or use an existing one. This tutorial uses the Nginx Ingress controller for routing the traffic. Minikube (v0.14 and above) comes with the Nginx setup as an addon which you will need to enable manually:
+
+```bash
+minikube addons enable ingress
+```
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: I suggest adding the following blockquote:

> Please note that it may take few minutes for Minikube to download and install Nginx as an ingress.

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+
+## Deploying a Laravel Application to Minikube
+
+As I've afore-mentioned, I will be showing you how to deploy a straightforward and stateless Laravel application to Kubernetes. I aim to detail the steps involved in accomplishing this while explaining why specific actions are required. Furthermore, I will show you how to quickly scale the application and also make it available on a named host using an Ingress controller.
+
+You can run Kubernetes on several cloud hosting providers such as Google Cloud Engine and Amazon Web Services. In this tutorial, I will run the application on Minikube, a tool that makes it easy to run Kubernetes locally.
+
+Minikube runs on a single node. You can think of a node as a virtual machine (VM). Inside a node (VM), you have the Docker daemon where you can run Docker containers. You can launch as many containers as you wish, as long as you have enough system resources such as memory and CPU. In this particular tutorial, you have one single node, but you're starting three replicas of this node.
```
[the remark this answers is not in the document]
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: Which "You"?

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+If you're still wondering why someone can benefit from using Kubernetes, the answer is simple. Kubernetes makes it easier to set up and manage as many clusters as required across multiple projects.
+
+## Deploying a Laravel Application to Minikube
+
+As I've afore-mentioned, I will be showing you how to deploy a straightforward and stateless Laravel application to Kubernetes. I aim to detail the steps involved in accomplishing this while explaining why specific actions are required. Furthermore, I will show you how to quickly scale the application and also make it available on a named host using an Ingress controller.
+
+You can run Kubernetes on several cloud hosting providers such as Google Cloud Engine and Amazon Web Services. In this tutorial, I will run the application on Minikube, a tool that makes it easy to run Kubernetes locally.
```
[the remark this answers is not in the document]
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: The "I" is intentional but "You" fits too. I'll update it.

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+
+Minikube runs on a single node. You can think of a node as a virtual machine (VM). Inside a node (VM), you have the Docker daemon where you can run Docker containers. You can launch as many containers as you wish, as long as you have enough system resources such as memory and CPU. In this particular tutorial, you have one single node, but you're starting three replicas of this node.
+Which translated from Kubernetes jargon becomes: You have 1 VM with 3 Docker containers running on it.
```
[the remark this answers is not in the document]
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: Which "You"?

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+
+It extends a PHP7 base image, installs some system dependencies including composer and the standard PHP extensions required by Laravel. It then copies the application files to a working directory and installs the application's dependencies via composer. It also runs a couple of `artisan` commands,  last of which serves the application using the built-in PHP web server. In the end, it exposes port `8181` to the host machine.
+  
+To build the local Docker image:
+
+```bash
+cd /to/your/project/directory
+eval $(minikube docker-env)
+docker build -t yourname/laravel-kubernetes-demo .
+```
```
[the remark this answers is not in the document]
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: 👍

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (danielepolencic) [reply _blog/deploying-laravel-to-kubernetes/index.md]: Which translated from Kubernetes jargon becomes: **You** have 1 VM

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+
+__Deploying the image__
+
+Now that the application's image is built and available locally you can go ahead with deploying it.
```
[the remark this answers is not in the document]
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: Should I add it or the _fyi_ is for me? Sorry, it just different on windows as a host machine.

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+kubectl config use-context minikube
+```
+
+then you can deploy the container image:
+
+```bash
+kubectl run laravel-kubernetes-demo --image=yourname/laravel-kubernetes-demo 
+--port=8181 --image-pull-policy=IfNotPresent
+```
+The above command tells Kubectl to run our demo application from the Docker image while making port 8181 available for listening. The last parameter of the command simply asks Kubectl to not pull the image from a registry such as Docker Hub if it exists locally which in this case it does. Do note that you still need to be logged on to Docker's so that Kubectl can check if the image is up to date.
```
[the remark this answers is not in the document]
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: I agree we need to be consistent. Mainly with using Caps or not for terms such as Pods, Nodes etc.. but maybe not ticks?

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 1 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: Should be _the application_. I'll fix it.

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 2 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: 👍

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 3 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: I agree with the text change. The numbers are correct though :) 1 + 2 __more__ = three 😄

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 4 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: Perfect 👍 I like it.

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 5 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: 😊

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 6 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: Okay. I'll drop the sentence because I didn't want to dedicate another paragraph to explain edge routers as it is a little out of context and too much info for the first time users.

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 7 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: I know. I didn't want to confuse the reader with _service_ as per previous PR's comments. I'll refactor and then you can let me know if it is confusing from an external point of view.

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 8 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: 👍

Comment (danielepolencic) [reply _blog/deploying-laravel-to-kubernetes/index.md]: Fair point! Let me explain what happens when you enable the add-on.

Under the bone, minikube is creating a deployment for Nginx. Same as the Laravel application. The Ingress is just another Pod with Nginx that acts as a reverse proxy. Nginx is packaged as a Docker container, exactly like your Laravel app.

So it may take a while to download the container and start the deployment.

In your case, I suspect the Docker image is cached locally. You can try and delete the cluster with:

```bash
$ minikube delete
```

and restart it with:

```bash
$ minikube start
```

If you enable the add-on you can quickly switch to another terminal and notice that something is happening with the pods:

```bash
$ kubectl get pods --all-namespaces
```

This part is very confusing, let me know if you want me to dive into more details.

**Comment by danielepolencic** [comment]
I like the article because it's very entry level. You've done a better job than I did in my workshop when it comes to introducing the basics.

I'm going to borrow some of the ideas that you have here. In particular, I liked:

- Deploying the application without using yaml. You introduced yaml only with the Ingress. I think there's an opportunity to explain why there're both, but I think this is probably not the place for that.
- I feel like the article is blurring the line between a Pod and the application/container. While this is technically not correct (a Pod can have multiple containers) I think it's the right thing to do. As a newbie I don't care how many containers are in a Pod, I just want to deploy the application.

A couple of notes, mostly for myself I think:

- [ ] Most of the people are visual learners. I think there're some concepts that could be reaffirmed using diagrams or images
- [ ] Missing intro image
- [ ] Ordering of blog posts in the /blog is not correct
- [ ] It would be nice to have links to share the article as a sidebar or at the bottom of the post

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (danielepolencic) [reply _blog/deploying-laravel-to-kubernetes/index.md]: I think for resources such as Pods and Nodes is fine. They're capitalised when you write resources in the yaml.
The ticks are usually for code or commands. Example: `kubectl`. I could write: `kubectl --version`. What do you think?

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 3 above)
Comment (danielepolencic) [reply _blog/deploying-laravel-to-kubernetes/index.md]: OH LOL. SORRY.

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+To follow with this demonstration, you will need the following installed on your local system:
+
+
+1) <a href="https://docs.docker.com/install/" target="_blank" rel="noopener">Docker</a>
+
+2) <a href="https://kubernetes.io/docs/tasks/tools/install-kubectl/" target="_blank" rel="noopener">Kubectl</a>
+
+3) <a href="https://github.com/kubernetes/minikube/releases" target="_blank" rel="noopener">Minikube</a>
+
+_Are you having problems installing and running these applications on Windows? LearnK8s will be publishing a tutorial on how to accomplish this very soon. Subscribe and be notified as soon as new articles get published._
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: Perhaps we could have this as a blockquote?

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+FROM php:7
+RUN apt-get update -y && apt-get install -y openssl zip unzip git
+RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer
+RUN docker-php-ext-install pdo mbstring
+WORKDIR /app
+COPY . /app
+RUN composer install
+
+RUN php artisan key:generate
+RUN php artisan serve --host=0.0.0.0 --port=8181
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: this is a CMD not RUN. When the container starts (at runtime) it executes the CMD command. RUN commands are executed at build time.

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+```bash
+kubectl get deployment
+
+NAME                      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
+laravel-kubernetes-demo   3         3         3            3           59m
+```
+
+You can also see this in the Dashboard under Pods or in the Service detail screen.
+
+Now the application's deployment services are running on three
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: I think this one is ambiguous to me. What about:

> Now you're running three instances of the applications using 3 Pods.

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: @danielepolencic I'm not understanding the question 🤔 Are you referring to every time we use kubectl as an application or when we execute the command? As for the former, I don't think we should and vice-versa for the latter. Maybe you also mean that when the command is one line, I should use single ticks as opposed to the block?

[the remark this answers is not in the document]
Comment (danielepolencic) [reply _blog/deploying-laravel-to-kubernetes/index.md]: [If you look at the definition of Pod in yaml](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#pod-v1-core) you will notice that it's capitalised: `kind: Pod`.
So it makes sense to keep it consistent.
kubectl is a binary in the shell. "Kubectl" with the capital K is not a valid command. Even in the official documentation kubectl is always lowercase: https://kubernetes.io/docs/reference/kubectl/overview/

[the remark this answers is not in the document]
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: I agree with you but in this case, I'm referring to Kubectl as the CLI application and not the command. Should I still include the ticks and/or write it all in lowercase?

[the remark this answers is not in the document]
Comment (danielepolencic) [reply _blog/deploying-laravel-to-kubernetes/index.md]: In the official documentation is referred like that. If you would rather have it capitalised, keep it that way 👍

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 8 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: Yepp, I had the Nginx image cached. I verified this by deleting Minikube and reinstalling it.

**Comment by danielepolencic** [comment]
@keithmifsud I'm already working on the diagrams ;)
I'm also getting the image too. You can leave those to me.

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: I'll change it so that it matches the docs :)

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 15 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: 👍

File: _blog/deploying-laravel-to-kubernetes/index.md (hunk 16 above)
Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: @danielepolencic This is confusing for me, mainly because both RUN & CMD worked. I had CMD before and changed it to RUN. Which ones should I change and to what please?

File: _blog/deploying-laravel-to-kubernetes/index.md [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+  name: laravel-kubernetes-demo-ingress
+  annotations:
+    ingress.kubernetes.io/rewrite-target: /
+spec:
+  backend:
+    serviceName: default-http-server
+    servicePort: 80
+  rules:
+  - host: laravel-kubernetes.demo
+  - http:
```
Comment (danielepolencic) [inline _blog/deploying-laravel-to-kubernetes/index.md]: there's an error. should be `http` without the `-`. `host` & `http` belong to the same item.

Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: 🤔 This works for me. Should I write it like this ?:

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: laravel-kubernetes-demo-ingress
  annotations:
    ingress.kubernetes.io/rewrite-target: /
spec:
  backend:
    serviceName: default-http-server
    servicePort: 80
  rules:
  - http:
      host: laravel-kubernetes.demo
      paths:
      - path: /
        backend:
          serviceName: laravel-kubernetes-demo
          servicePort: 8181
```

Comment (danielepolencic) [reply _blog/deploying-laravel-to-kubernetes/index.md]: I think so: https://kubernetes.io/docs/concepts/services-networking/ingress/#the-ingress-resource

Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: @danielepolencic I tried it and it doesn't work. The example link you've included does not have the _host_ key under http. The old way works.
ingress_host_issue

Comment (danielepolencic) [reply _blog/deploying-laravel-to-kubernetes/index.md]: weird. Let's leave it like this then. sorry.

Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: I don't see it as weird because the host key is to set up a host on nginx (http) and not the actual Ingress. I assume we're still waiting for the images? You or I will also need to change the article date before merging.

Comment (danielepolencic) [reply _blog/deploying-laravel-to-kubernetes/index.md]: We have the image. It needs some tweaking. I also have the diagrams too. I'm a little bit swamped atm. As soon as I have more capacity I'll finish this off. Sorry for the delay.

Comment (keithmifsud) [reply _blog/deploying-laravel-to-kubernetes/index.md]: No worries. Once published I can also help with sharing the article 😄
